### ISSUE TITLE
conditional fields and reset

### ISSUE THREAD

**Issue by lobopl** [react-hook-form/react-hook-form#1041]
**Describe the question?**
I have quite big form (60+ fields) that has a lot of conditional fields and it work great if user is providing all data by hand, but when i load data from database and use reset i have issue with values that stay in conditional inputs and some fields don't disappear after user action

**To Reproduce**
1. Open first sanbox link after 1 second all fields will get value and start to be visible,
2. Uncheck field-1, field-2 and field-3 will disappear but field-4 will stay visible
3. Check field-1 again field-2 and field-3 will appear with default value

**Codesandbox link**
https://codesandbox.io/s/react-hook-form-watch-z7t5g

I tried to fill form using setValue on each field but it took over 15 seconds is there a way to fill fields and have default values as empty?

**Comment by bluebill1049** [comment]
The example is hard to understand...

**Comment by bluebill1049** [comment]
Please supply each scenario below with an example:

1. when i load data from database and use reset i have issue with values that stay in conditional inputs 

2. and some fields don't disappear after user action

**Comment by lobopl** [comment]
Maybe i will try to rephrase it
in my example i have 4 fields (field-1, field-2, field-3, field-4)
field-2, field-3, field-4 watch some fields to be visible (field-2 and field-3 watch field 1, field-4 watch field-3)

1. When you open my example all fields will get values in async function (setTimeout simulating ajax request), values are set using reset. Now all fields have value and all fields they are watching also have value so they render, so far it is working as expected.
2. Now uncheck field-1, after this action field-2 and field-3 that are watching field-1 will disappear but field-4 that is watching field-3 is still visible
3. If i check again field-1, field-2 and field-3 will appear again and they will again have value true that was provided in reset


What i want to achieve:
1. same as above
2. uncheck field-1 will hide all other fields
3. checking field-1 again will show field-2 with empty value and field-3 as unchecked, field-4 will remain hidden
4. checking field-3 will show field-4 with empty value

basically i want to set all fields value from backend in nested conditional fields, but i don't want to have default values that will fill conditional inputs when they appear from user action.

**Comment by bluebill1049** [comment]
> basically i want to set all fields value from backend in nested conditional fields, but i don't want to have default values that will fill conditional inputs when they appear from user action.

Try to make conditional input with `display: none` instead of unmount them.

**Comment by lobopl** [comment]
this will not work for 2 reasons
1. Values will stay same as initially loaded or if user changed them they will be same as changed ones and not empty as i want
2. if field is required (one that is hidden) form will see it as incorrect, imagine something like first checkbox is optional, but conditional one is required

**Comment by bluebill1049** [comment]
valid points above, please consider to upload two seperate example and keep them simple as what i suggested above, i am bit tried today and tried to make seperate CSB, so if you could help me/us would be nice.

**Comment by lobopl** [comment]
this is example how i want this to work:
https://codesandbox.io/s/react-hook-form-watch-7xn9x
and i don't want to use setValue in this timeout because i have 60+ fields and in this kind of action form renders 60+ times and it take 15-20 seconds to render

**Comment by lobopl** [comment]
Is there a way to set values for all fields at once other than reset and multiple setValue? I don't want default values to be applied when field is removed and then appear again because something changed.

**Comment by bluebill1049** [comment]
Is this what you after? https://codesandbox.io/s/happy-gauss-en8jy

**Comment by lobopl** [comment]
not really this is basically same concept as in  my "problematic form using reset", basically what i want to achieve is setting big form with conditional fields but without using reset because it sets default values and if field dissapear and then appear again instead of being empty it gets this default value

https://codesandbox.io/s/react-hook-form-watch-7xn9x
first form click field-4 has value from server 2 if you click check and uncheck field-3 field-4 will disappear then appear and it will have no value inside this is what i want

second form if you do exactly the same as above field-4 will reset value to default from server 2

I cannot really use approach from first form because my form has 60+ fields and setValue trigger render 60+ times and it is quite slow (15-20 seconds)

**Comment by bluebill1049** [comment]
> because it sets default values and if field dissapear and then appear again instead of being empty it gets this default value

what do you mean by that? did you see my demo and try to type '123' in the input?

**Comment by lobopl** [comment]
yes and it presented "with reset value", but in my use case i wanted something like this
1. I load data from backend (same as in your example but 123 instead of 1234)
2. all fields are visible
3. i change second text input to something like "test"
4. I change first input to 1234 second text input disappear
5. I change first input to 123 second text input appear with nothing as value not "with reset value" and not "test"

so if field disappear values in that input are lost, and all fields watching that field will get information that there is no value currently. Basically i want to set all field values at once but not defaultValues, and i want form to render only once and not each time for running setValue

**Comment by bluebill1049** [comment]
React Hook Form embraced uncontrolled input and works just like uncontrolled.

`{show && <input name="test" defaultValue="test" />}`

`reset` API default values is the same as change `defaultValue` of that input, which means when you 

1. type 123, it will show 'test' 
2. type 1234 get removed
3. type 123, it will show 'test' or newly updated default value

if uncheck an item need to clear a field, then you can add `onChange` event and set the input to empty.

**Comment by lobopl** [comment]
Yes i could do that but it is quite complicated to manage big nested form with dynamic fields. And in addition i would need to keep state for many fields and change logic from watching values in child fields that depend on other fields to pushers in parents. And now those "parent" fields would need to have knowledge about childs. Right now my temporary solution is resetting only part of form containing fields that are not conditional and then use setValue on each field that is conditional and have value from api, its slow because there is up to 30 conditional fields and 30 not conditional (so 10-15 seconds for render 30+ renders). Would be nice if there was one method setValues allowing setting multiple values at once and working similar to reset, but without setting defaultvalues

**Comment by bluebill1049** [comment]
`setValue` support nest/multiple objects is the next feature (on the todo list).

**Comment by lobopl** [comment]
May i know in what time range it would be created or it isn't yet planned?

**Comment by bluebill1049** [comment]
the next 2 weeks. it's on the project board.

**Comment by bluebill1049** [comment]
actually not sure if the feature is relevant to your request: 

```
set('name', [{ test: '123' }])
set('name[0]', { test: '123' })
```

because the feature is about support nest objects and multiple array item values.

**Comment by lobopl** [comment]
Sounds great thanks.
